README.md
# sliding_switch

Sliding Switch - A simple switch widget. It can be fully customized with desired width, colors, text etc. It also maintains selection state.

## Live Demo (Codeapprun)

[![codeapprun.io](https://storage.googleapis.com/s3.codeapprun.io/assets/badge.svg)](https://codeapprun.io/sachindarde/widget/MgkxPPDfch)
## Getting Started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  sliding_switch: "latest"
```

Import it:

```dart
import 'package:sliding_switch/sliding_switch.dart';
```

## Usage Examples

### Minimum sliding switch configuration

```dart
SlidingSwitch(
 value: false,
 width: 250,
 onChanged: (bool value) {
   print(value);
 },
)
```

![sliding switch](https://storage.googleapis.com/s3.codeapprun.io/assets/sliding_gif.gif)

### sliding switch other configurations

```dart
SlidingSwitch(
 value: false,
 width: 250,
 onChanged: (bool value) {
   print(value);
 },
 height : 55,
 animationDuration : const Duration(milliseconds: 400),
 onTap:(){},
 onDoubleTap:(){},
 onSwipe:(){},
 textOff : "Female",
 textOn : "Male",
 iconOff: Icons.human-female,
 iconOn: Icons.human-male,
 contentSize: 17,
 colorOn : const Color(0xffdc6c73),
 colorOff : const Color(0xff6682c0),
 background : const Color(0xffe4e5eb),
 buttonColor : const Color(0xfff7f5f7),
 inactiveColor : const Color(0xff636f7b),
),
```

If iconOn or iconOff are not null, they will be displayed in lieu of the respective textOn and textOff.  The values for textOn and textOff are used as semantic labels for the icons for accessibility.  The icons in the example are not in the default material icon set, but are from [Material Design Icons] (https://materialdesignicons.com/), available on pub.dev [here](https://pub.dev/packages/material_design_icons_flutter).

contentSize drives the height of the text or icons.

Licensed Under the [MIT License](LICENSE).

## Inspiration

[Jitu Raut](https://dribbble.com/shots/6190542-BMI-Calculator-Interaction)


CHANGELOG.md
## [0.0.1] - [2020/08/16]
    * Custom Color
    * Selection state
    * Sizing
    * Custom Text
## [0.0.2] - [2020/08/19]
    * On change method was being called on load fix
    
## [0.0.3] - [2020/08/23]
    * Minor changes
## [1.0.0] - [2020/08/23]
    * Added Null Safety
    * Removed conflict

lib/sliding_switch.dart
import 'package:flutter/material.dart';

class SlidingSwitch extends StatefulWidget {
  final double height;
  final ValueChanged<bool> onChanged;
  final double width;
  final bool value;
  final String textOff;
  final String textOn;
  final IconData iconOff;
  final IconData iconOn;
  final double contentSize;
  final Duration animationDuration;
  final Color colorOn;
  final Color colorOff;
  final Color background;
  final Color buttonColor;
  final Color inactiveColor;
  final Function onTap;
  final Function onDoubleTap;
  final Function onSwipe;

  const SlidingSwitch({
    required this.value,
    required this.onChanged,
    this.height = 55,
    this.width = 250,
    this.animationDuration = const Duration(milliseconds: 400),
    required this.onTap,
    required this.onDoubleTap,
    required this.onSwipe,
    this.textOff = "Off",
    this.textOn = "On",
    this.iconOff,
    this.iconOn,
    this.contentSize = 17,
    this.colorOn = const Color(0xffdc6c73),
    this.colorOff = const Color(0xff6682c0),
    this.background = const Color(0xffe4e5eb),
    this.buttonColor = const Color(0xfff7f5f7),
    this.inactiveColor = const Color(0xff636f7b),
  }) : super();
  @override
  _SlidingSwitch createState() => _SlidingSwitch();
}

class _SlidingSwitch extends State<SlidingSwitch>
    with SingleTickerProviderStateMixin {
  late AnimationController animationController;
  late Animation<double> animation;
  double value = 0.0;

  late bool turnState;

  @override
  void dispose() {
    animationController.dispose();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();
    animationController = AnimationController(
        vsync: this,
        lowerBound: 0.0,
        upperBound: 1.0,
        duration: widget.animationDuration);
    animation =
        CurvedAnimation(parent: animationController, curve: Curves.easeInOut);
    animationController.addListener(() {
      setState(() {
        value = animation.value;
      });
    });
    turnState = widget.value;
    _determine();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
        onDoubleTap: () {
          _action();
          if (widget.onDoubleTap != null) widget.onDoubleTap();
        },
        onTap: () {
          _action();
          if (widget.onTap != null) widget.onTap();
        },
        onPanEnd: (details) {
          _action();
          if (widget.onSwipe != null) widget.onSwipe();
        },
        child: Container(
          height: widget.height,
          width: widget.width,
          decoration: BoxDecoration(
              color: widget.background,
              borderRadius: BorderRadius.circular(50)),
          padding: EdgeInsets.all(2),
          child: Stack(children: <Widget>[
            Transform.translate(
                offset: Offset(((widget.width * 0.5) * value - (2 * value)), 0),
                child: Container(
                  height: widget.height,
                  width: widget.width * 0.5 - 4,
                  decoration: BoxDecoration(
                      borderRadius: BorderRadius.all(Radius.circular(50.0)),
                      color: widget.buttonColor,
                      boxShadow: [
                        new BoxShadow(
                          color: Colors.black.withOpacity(0.2),
                          offset: Offset(0, 10),
                          blurRadius: 20.0,
                        ),
                      ]),
                )),
            Row(
              children: [
                Expanded(
                  child: Center(
                    child: widget.iconOff == null
                      ? Text(
                          widget.textOff,
                          style: TextStyle(
                              color: turnState
                                  ? widget.inactiveColor
                                  : widget.colorOff,
                              fontSize: widget.contentSize,
                              fontWeight: FontWeight.w600),
                        )
                      : Icon(widget.iconOff,
                          semanticLabel: widget.textOff,
                          size: widget.contentSize,
                          color: turnState
                              ? widget.inactiveColor
                              : widget.colorOff,
                      ),
                  ),
                ),
                Expanded(
                  child: Center(
                    child: widget.iconOn == null
                    ? Text(
                        widget.textOn,
                        style: TextStyle(
                            color:
                                turnState ? widget.colorOn : widget.inactiveColor,
                            fontSize: widget.contentSize,
                            fontWeight: FontWeight.w600),
                      )
                    : Icon(widget.iconOn,
                          semanticLabel: widget.textOn,
                          size: widget.contentSize,
                          color: turnState
                              ? widget.colorOn
                              : widget.inactiveColor,
                      ),
                  ),
                )
              ],
            )
          ]),
        ));
  }

  _action() {
    _determine(changeState: true);
  }

  _determine({bool changeState = false}) {
    setState(() {
      if (changeState) turnState = !turnState;
      (turnState)
          ? animationController.forward()
          : animationController.reverse();
      if (changeState) widget.onChanged(turnState);
    });
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:sliding_switch/sliding_switch.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Welcome to Flutter',
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        backgroundColor: Colors.white,
        body: Center(
          child: SlidingSwitch(
            value: false,
            width: 250,
            onChanged: (bool value) {
              print(value);
            },
            height: 55,
            animationDuration: const Duration(milliseconds: 400),
            onTap: () {},
            onDoubleTap: () {},
            onSwipe: () {},
            textOff: "Female",
            textOn: "Male",
            colorOn: const Color(0xffdc6c73),
            colorOff: const Color(0xff6682c0),
            background: const Color(0xffe4e5eb),
            buttonColor: const Color(0xfff7f5f7),
            inactiveColor: const Color(0xff636f7b),
          ),
        ),
      ),
    );
  }
}
